### PR TITLE
Use Github tags if present

### PR DESCRIPTION
**Pull request by adam-hart**
If a release is about to be created, and it has a tag already defined on Github, use that one for a release instead of creating a new one in Samson.

/cc @zendesk/samson @zendesk/sustaining 

### Risks
- Medium:  Could cause release numbers to go all sorts of wrong.

### CODE REVIEW

File: app/models/release.rb [hunk 1]
```diff
@@ -56,9 +56,15 @@ def assign_release_number
     # release-number-from-ci plugin.
     return if number != DEFAULT_RELEASE_NUMBER && number.present?
 
-    latest_release_number = project.releases.last.try(:number) || "0"
+    raise "Unable to auto bump version" unless self.number = next_release_number
+  end
+
+  def next_release_number
+    # If Github has a version tagged for this commit, use that version instead of ours
+    latest_samson_version = Gem::Version.new(project.releases.last.try(:number) || "0").to_s
```
Comment (grosser) [inline app/models/release.rb]: can use fancy new `. last&.number`

File: app/models/release.rb [hunk 2]
```diff
@@ -56,9 +56,15 @@ def assign_release_number
     # release-number-from-ci plugin.
     return if number != DEFAULT_RELEASE_NUMBER && number.present?
 
-    latest_release_number = project.releases.last.try(:number) || "0"
+    raise "Unable to auto bump version" unless self.number = next_release_number
```
Comment (grosser) [inline app/models/release.rb]: no need for self.

Comment (grosser) [inline app/models/release.rb]: add an error instead of raising ?

File: app/models/release.rb [hunk 3]
```diff
@@ -56,9 +56,15 @@ def assign_release_number
...
+  end
+
+  def next_release_number
+    # If Github has a version tagged for this commit, use that version instead of ours
+    latest_samson_version = Gem::Version.new(project.releases.last.try(:number) || "0").to_s
+    latest_github_version = Gem::Version.new(project.repository.tag_from_ref(commit)).to_s
 
-    raise "Unable to auto bump version" unless self.number = latest_release_number.dup.sub!(/\d+$/) { |d| d.to_i + 1 }
+    latest_github_version.present? ? latest_github_version : latest_samson_version.to_s.dup.sub!(/\d+$/) { |d| d.to_i + 1 }
```
Comment (grosser) [inline app/models/release.rb]: what about that "and is greater then last" ?

File: app/models/release.rb (hunk 2 above)
Comment (sandlerr) [inline app/models/release.rb]: shouldn't redundant `self` be a rubocop ci error?

**Comment by grosser** [comment]
yep

File: app/models/release.rb (hunk 2 above)
Comment (adam-hart) [inline app/models/release.rb]: Something is funny here.  I'm not sure if it needs an `attr_writer`, but if I remove the `self`, it doesn't stick.

File: app/models/release.rb [hunk 4]
```diff
@@ -56,9 +56,19 @@ def assign_release_number
     # release-number-from-ci plugin.
     return if number != DEFAULT_RELEASE_NUMBER && number.present?
 
-    latest_release_number = project.releases.last.try(:number) || "0"
+    errors.add :number, "Unable to auto bump version" unless self.number = next_release_number
+  end
+
+  def next_release_number
```
Comment (grosser) [inline app/models/release.rb]: private ?

File: test/models/release_test.rb [hunk 5]
```diff
@@ -10,6 +10,10 @@
   let(:commit) { "abcde" * 8 }
 
   describe "create" do
+    before do
+      GitRepository.any_instance.stubs(:tag_from_ref).returns("")
```
Comment (grosser) [inline test/models/release_test.rb]: why do we need to stub this ? ... too slow ?

Comment (adam-hart) [inline test/models/release_test.rb]: Yeah, this request seems to fail left and right on Travis.

**Pull request merged**